### PR TITLE
New Process class

### PR DESCRIPTION
**Pull request by neikeq**
### DO NOT MERGE GUARD

This class allows spawning child processes in a non-blocking way while, unlike `OS::execute()`, allowing to keep track of its lifetime and doing IO operations.

The interface is heavily inspired on QProcess and a bit on D's std.process (redirect flags and env).

Notes:
- This commit removes our old SIGCHLD handler (introduced by #7774), which would conflict with forkfd's one that is used by the Process class. OS_Unix::execute() now depends on forkfd, which is used instead of fork() if the condition `!p_blocking` is met.
- `OS_Unix::execute()` calls forkfd_nofd() instead of forkfd(). Unlike the latter, forkfd_nofd() does not create any file descriptor to track the child process. This allows forkfd to be used by `OS_Unix::execute()` without gdb notifying (unless a startup command is specified) of SIGPIPE signals, since `OS_Unix::execute()` would immediately close the returned file descriptor otherwise.

Tested on the following platforms:

- [x] x11 (Ubuntu 16.04 64bit)
- [ ] server
- [x] windows (Windows 7 64bit)
- [ ] osx
- [ ] iphone
- [ ] android
- [ ] haiku
- [ ] bb10
- [ ] uwp

I can't test on the unchecked platforms, I don't have access to them (well, android, but...).

Example: https://gist.github.com/neikeq/5b3ab0461733d1100561b0ba227a19e0

Docs will come later.

### CODE REVIEW

File: core/os/process.cpp [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+	ClassDB::bind_method(D_METHOD("get_cwd"), &Process::get_cwd);
+
+	ClassDB::bind_method(D_METHOD("get_state"), &Process::get_state);
+	ClassDB::bind_method(D_METHOD("get_exit_status"), &Process::get_exit_status);
+	ClassDB::bind_method(D_METHOD("get_exit_code"), &Process::get_exit_code);
+
+	ClassDB::bind_method(D_METHOD("set_redirect_flags", "flags"), &Process::set_redirect_flags);
+	ClassDB::bind_method(D_METHOD("get_redirect_flags"), &Process::get_redirect_flags);
+
+	ClassDB::bind_method(D_METHOD("set_environent", "env", "override"), &Process::_set_environment_bind, DEFVAL(false));
```
Comment (opcon) [inline core/os/process.cpp]: I think `set_environent` should be `set_environment`?

`environent` occurs in a few other places as well.

Comment (neikeq) [inline core/os/process.cpp]: Indeed.

**Comment by akien-mga** [comment]
This is getting stale, what's the plan for it? :)

**Comment by neikeq** [comment]
I am closing this. It's unfortunate, but Mono has a unfriendly SIGCHLD handler. It catches all the processes that exited, not only the ones it spawned. It does not have signal chaining either. That makes this PR conflict with builds that use mono and spawn processes.
I thought about some hackish solutions, but none really convinced me and I am focused on other stuff right now. Maybe I'll pick this up in the future.

**Comment by Calinou** [comment]
> I am closing this. It's unfortunate, but Mono has a unfriendly SIGCHLD handler. It catches all the processes that exited, not only the ones it spawned. It does not have signal chaining either. That makes this PR conflict with builds that use mono and spawn processes. I thought about some hackish solutions, but none really convinced me and I am focused on other stuff right now. Maybe I'll pick this up in the future.

Out of curiosity, is this still an issue with .NET Core? I've started to look at non-blocking `OS.execute()` interaction since it would allow for implementing a feature in a more performant way (GPU clocks/temperature readouts via `nvidia-smi`).

**Comment by neikeq** [comment]
> Out of curiosity, is this still an issue with .NET Core?

The .NET Core runtime can be either CoreCLR or Mono. I would expect the Mono runtime to still have this problem. No idea about CoreCLR.

**Pull request closed without merging**